### PR TITLE
udisks plugin: fix memory management for GValue variables

### PR DESCRIPTION
**Pull request by monsta**
fixes build warnings as well

### CODE REVIEW

**Comment by info-cppsp** [comment]
For future reference, I would like to point out, that `g_value_unset()` calls are needed, so that the associated string variables (ids, model, dev) are freed.

Before they were freed with g_free() and the GValues were freed automatically at function end.

Now that I think about it, freeing them wasn't necessary before either.
In fact I don't think using `g_value_unset()` is even needed.

Please check.

**Comment by info-cppsp** [comment]
Also, this is close to being a duplicate of #65 ...
That was first...

**Comment by monsta** [comment]
Who would free the internal string pointers inside GValue then?

**Comment by info-cppsp** [comment]
https://github.com/mate-desktop/mate-sensors-applet/pull/65#issuecomment-358279570
I thought that it is the same here.
Now I am very confused.

The G_TYPE_VALUE used in the plugin is a boxed type.
https://developer.gnome.org/gobject/stable/gobject-Boxed-Types.html

The above G_TYPE_STRING clearly has the autofree at the end of the function.
It is in its type definition.

The boxed type doesn't have something like that.
That would mean that at least the boxed variables should be freed, like in the commit.

From glib/gobject/tests/boxed.c:
```c
static void
test_boxed_value (void)
{
  GValue value1 = G_VALUE_INIT;
  GValue *value2;
  GValue value = G_VALUE_INIT;

  g_value_init (&value, G_TYPE_VALUE);
  g_assert (G_VALUE_HOLDS_BOXED (&value));

  g_value_init (&value1, G_TYPE_INT);
  g_value_set_int (&value1, 26);

  g_value_set_static_boxed (&value, &value1);

  value2 = g_value_get_boxed (&value);
  g_assert (value2 == &value1);

  value2 = g_value_dup_boxed (&value);
  g_assert (value2 != &value1);
  g_assert (G_VALUE_HOLDS_INT (value2));
  g_assert_cmpint (g_value_get_int (value2), ==, 26);
  g_boxed_free (G_TYPE_VALUE, value2);

  g_value_unset (&value);
}
```

Here 'value' is the boxed type. It gets freed with g_value_unset().
I think 'value1' gets freed at function end, not with g_value_unset().
[boxed.c.txt](https://github.com/mate-desktop/mate-sensors-applet/files/1660043/boxed.c.txt)
(Rename and) compile with
```
gcc -Wall -g boxed.c -o boxed `pkg-config --cflags --libs glib-2.0 gobject-2.0`
```
-> 'value1' is not freed by g_value_unset().

[boxed.c.v2.txt](https://github.com/mate-desktop/mate-sensors-applet/files/1660295/boxed.c.v2.txt)

```c
static void
test_boxed_value (void)
{
  GValue value1 = G_VALUE_INIT;
  GValue *value2;
  GValue value = G_VALUE_INIT;

  g_value_init (&value, G_TYPE_VALUE);
  g_assert (G_VALUE_HOLDS_BOXED (&value));

  g_value_init (&value1, G_TYPE_INT);
  g_value_set_int (&value1, 26);

  /* set internal data pointer of value to point to the data of value1 */
  g_value_set_static_boxed (&value, &value1);

  /* get where the internal data pointer of value points to */
  value2 = g_value_get_boxed (&value);
  /* this is TRUE */
  g_assert (value2 == &value1);

  /* free value - this doesn't free the data!
   * bc the GValue structure itself is freed, which has only a pointer to the real data of value1
   * comment out g_value_unset (value2); line to see */
  g_value_unset (&value);

  /* free value2
   * as value2 points to value1, value1 is freed
   * (same as g_value_unset (&value1);)
   * in this case with its data, bc it is of type G_TYPE_INT
   * (the data is in the GValue - for a string the GValue holds only a pointer) */
  g_value_unset (value2);

  /* this fails, bc value1 is already freed */
  g_assert (G_VALUE_HOLDS_INT (&value1));
  g_assert_cmpint (g_value_get_int (&value1), ==, 26);
printf("my int %d", g_value_get_int (&value1));
}
```

What does this mean for the udisks plugin?
AFAI can tell, the GValues model_v, dev_v, ids_v - their structure is freed at the end of the function.
You don't have to bother with them.
BUT their data is not freed! Not even with g_value_unset(), as the GValues in the plugin only hold pointers to the data, which they don't know how to free.

SO
```
            g_value_unset(&ids_v);
            g_value_unset(&model_v);
            g_value_unset(&dev_v);
```
is useless.

```
            g_strfreev(ids);
            g_free(model);
            g_free(dev);
```
is still the way to go.

Please double check.

If this comment is true, I was obviously wrong in [this one](https://github.com/mate-desktop/mate-sensors-applet/pull/70#issuecomment-359813939).
I also caught monsta trying to introduce a memory leak. :P

**Comment by monsta** [comment]
> BUT their data is not freed! Not even with g_value_unset(), as the GValues in the plugin only hold pointers to the data, which they don't know how to free.

I'll check what's up with the boxed type, but as for the internal string pointers, your example from https://github.com/mate-desktop/mate-sensors-applet/pull/65#issuecomment-358279570 already shows how they're freed when `g_value_unset` is called.

**Comment by muktupavels** [comment]
@info-cppsp that function does not return newly allocated data! so you should not free return value. there is reason why return value is const. There is g_value_dup_string that returns newly allocated string that should be freed.

**Comment by muktupavels** [comment]
> Here 'value' is the boxed type. It gets freed with g_value_unset().
> I think 'value1' gets freed at function end, not with g_value_unset().

`value1` does not allocate anything that should be freed.

**Comment by muktupavels** [comment]
Why `GValue` is used in first place?

Since returned type is `G_TYPE_VALUE` correct thing probably would require to check that `GValue` can hold `G_TYPE_STRING` before calling `g_value_get_string`, but if it is known that returned data is string then why not do this:

```
gchar *model;
dbus_g_proxy_call (sensor_proxy, "Get", NULL,
                   G_TYPE_STRING, UDISKS_BUS_NAME,
                   G_TYPE_STRING, "DriveModel",
                   G_TYPE_INVALID,
                   G_TYPE_STRING, &model,
                   G_TYPE_INVALID);

/* do something with model */
g_free (model);
```

**Comment by info-cppsp** [comment]
@monsta 
Yes. For the G_TYPE_STRING, the internal data, the string, is freed bc there is a free function given in the type definition.

This is the type definition for the G_TYPE_VALUE (the boxed):
glib/gobject/gboxed.c
```c
void
_g_boxed_type_init (void)
{
  const GTypeInfo info = {
    0,                          /* class_size */
    NULL,                       /* base_init */
    NULL,                       /* base_destroy */
    NULL,                       /* class_init */
    NULL,                       /* class_destroy */
    NULL,                       /* class_data */
    0,                          /* instance_size */
    0,                          /* n_preallocs */
    NULL,                       /* instance_init */
    NULL,                       /* value_table */
  };
  const GTypeFundamentalInfo finfo = { G_TYPE_FLAG_DERIVABLE, };
  GType type;

  /* G_TYPE_BOXED
   */
  type = g_type_register_fundamental (G_TYPE_BOXED, g_intern_static_string ("GBoxed"), &info, &finfo,
				      G_TYPE_FLAG_ABSTRACT | G_TYPE_FLAG_VALUE_ABSTRACT);
  g_assert (type == G_TYPE_BOXED);
}
```

No free function.

g_value_unset() would call the free function.
gvalue.c
```c
/**
 * g_value_unset:
 * @value: An initialized #GValue structure.
 *
 * Clears the current value in @value (if any) and "unsets" the type,
 * this releases all resources associated with this GValue. An unset
 * value is the same as an uninitialized (zero-filled) #GValue
 * structure.
 */
void
g_value_unset (GValue *value)
{
  GTypeValueTable *value_table;
  
  if (value->g_type == 0)
    return;

  g_return_if_fail (G_IS_VALUE (value));

  value_table = g_type_value_table_peek (G_VALUE_TYPE (value));

  if (value_table->value_free)
    value_table->value_free (value);
  memset (value, 0, sizeof (*value));
}
```

ALL 3 ARE THE BOXED TYPE!

@muktupavels 
1. if you mean g_value_get_string(), you are correct. BUT bc of the boxed type GValue, the string is not freed otherwise, so someone must free it.

2. I'm sorry. I meant that the GValue structures are freed.
TBH I'm not sure about the `g_value_unset (&value);` in the first code.
But that resets the GValue structure to an initial state.
At function end both value and value1 are freed - in this case with their data.
(value has no data and value1 has an int data, which is autofreed, bc it is a G_TYPE_INT)

3. 'Since returned type is G_TYPE_VALUE correct thing probably would require to check that GValue can hold G_TYPE_STRING before calling g_value_get_string'
You have a point there.
'if it is known that returned data is string'
It is not. At least:
screenshot at 2018-01-25 17 00 47
screenshot at 2018-01-25 17 01 50
screenshot at 2018-01-25 17 02 28

Look at the method name line. The return type is a GVariant.
Look at the udisks2 plugin to see how I have dealt with them.
Better would be to use G_TYPE_VARIANT then. (for model_v, dev_v, ids_v)

**Comment by info-cppsp** [comment]
@muktupavels 
3. I just remembered:
```c
/**
 * g_value_get_string:
 * @value: a valid #GValue of type %G_TYPE_STRING
 *
 * Get the contents of a %G_TYPE_STRING #GValue.
 *
 * Returns: string content of @value
 */
const gchar*
g_value_get_string (const GValue *value)
{
  g_return_val_if_fail (G_VALUE_HOLDS_STRING (value), NULL);
  
  return value->data[0].v_pointer;
}
```
So there is a check.

**Comment by muktupavels** [comment]
You don't free and/or modify things that are returned as const, period. You did not allocate it, you don't free it. It works because `dbus_g_proxy_call()` probably use `g_value_set_string`, but now lets imagine that it used `g_value_set_static_string()`. Now you would free static string that might be used elsewhere. Sooner or later it will crash because code will try to access freed data. You should use `g_value_unset()` and it will free string if needed.

`G_TYPE_VALUE` is abstract type. That is why I told that correct thing probably would be to check if GValue actually holds string before using `g_value_get_string().` In this case `GValue` is either `G_TYPE_STRING` or `G_TYPE_STRV` and both have free functions, no?

**Comment by muktupavels** [comment]
> 'Since returned type is G_TYPE_VALUE correct thing probably would require to check that GValue can hold G_TYPE_STRING before calling g_value_get_string'
> You have a point there.
> 'if it is known that returned data is string'
> It is not. At least:

You made me to look at interface and I do see that `DeviceFile` is string and `DeviceFileById` is array of string. Am pretty sure that code I posted would work and you dont need `GValue` and or `GVariant`

**Comment by muktupavels** [comment]
And for ids:
```
gchar **ids
dbus_g_proxy_call (sensor_proxy, "Get", NULL,
                   G_TYPE_STRING, UDISKS_BUS_NAME,
                   G_TYPE_STRING, "DeviceFileById",
                   G_TYPE_INVALID,
                   G_TYPE_STRV, &ids,
                   G_TYPE_INVALID);

/* do something with ids */
g_strfreev (ids);
```

**Comment by muktupavels** [comment]
@info-cppsp check this example - https://pastebin.com/yKr8UqKF to see what happens when you do something that you should not do. Same thing, but does correct thing - https://pastebin.com/8NAXHhr9

**Comment by info-cppsp** [comment]
@muktupavels 
1.
'You don't free and/or modify things that are returned as const, period.'
You don't say...
Why free the variables ids, model, dev in the first place??

'It works because dbus_g_proxy_call() probably use g_value_set_string'
We aren't really interested in theories about existing code... ;-)

'You should use g_value_unset() and it will free string if needed.'
I have just explained why I think that for the boxed type the pointed-to-data, in this case a string, will not be freed with g_value_unset().

'In this case GValue is either G_TYPE_STRING or G_TYPE_STRV and both have free functions, no?'
'ALL 3 ARE THE BOXED TYPE!' - to be more precise G_TYPE_VALUE
It is another matter, that in this case, like for the DriveModel the GValue model_v is going to be a G_TYPE_VALUE type, that holds a string value.
The type of model_v is however still G_TYPE_VALUE.
(So it still doesn't have a free function.)

2. Let's assume the people who wrote the udisks plugin weren't idiots.
Why haven't they used the G_TYPE_STRING ?

-----
OK, so let me tell you about dbus_g_proxy_call().
It calls dbus_g_proxy_end_call_internal().
http://maemo.org/api_refs/4.1/dbus-glib-0.74/dbus-gproxy_8c-source.html#l02214
[This](http://maemo.org/api_refs/4.1/dbus-glib-0.74/dbus-gproxy_8c-source.html#l02274) is the most interesting part.

Path a:
If we have a GVariant return type from DBus (we do) and if the user wants a G_TYPE_VALUE (we do), call _dbus_gvalue_demarshal_variant().
Demarshal stuff:
http://maemo.org/api_refs/4.0/dbus-glib/dbus-gvalue_8c-source.html
This gets messy, but I think _dbus_gvalue_demarshal_variant() essentially creates a G_TYPE_VALUE GValue from the GVariant returned from DBus.
If the GVariant holds a string, the string will be copied into the GValue with g_value_set_string().
Variables reply and msgiter are freed at the end of dbus_g_proxy_end_call_internal(), I think along with the original GVariant.

We get a GValue of type G_TYPE_VALUE, that holds our data.

Path b:
If we go the other way, like using G_TYPE_STRING instead of G_TYPE_VALUE in gbus_g_proxy_call(), we store the demarshaled GValue, it is already of the type we want (G_TYPE_STRING), into the return pointer with [_dbus_gvalue_store()](http://maemo.org/api_refs/4.0/dbus-glib/dbus-gvalue-utils_8c-source.html#l00074).
(Essentially getting the raw data.)
For G_TYPE_STRING
`*((gchar **) storage) = (char*) g_value_get_string (value);`
is used.

We get the raw data.

-----
In both cases the data needs to be freed.
(Well, it is not fully clear why in Path b the GValue doesn't autofree its data, but based on the [comment](http://maemo.org/api_refs/4.1/dbus-glib-0.74/dbus-gproxy_8c-source.html#l02300) it is like that...)

**Comment by muktupavels** [comment]
Ok, G_TYPE_STRING can not be used as I think, bet see this:
https://pastebin.com/6ymL1dmt

Returned value is `G_TYPE_STRING` not `G_TYPE_VALUE`...

**Comment by info-cppsp** [comment]
[test3.c.txt](https://github.com/mate-desktop/mate-sensors-applet/files/1665607/test3.c.txt)
I have wasted so much time with this already, I had to see this to the end.
I didn't want to, but fired up my old comp, and made a test.

Output:
```
model type name: gchararray
dev type name: gchararray
id type name: GStrv
model: Maxtor 6Y080L0
dev: /dev/sda
id: /dev/disk/by-id/ata-Maxtor_...

After freeing the GValues:
model type name: (null)
dev type name: (null)
id type name: (null)
model: �z�
dev: 0*�
id: 
Segmentation fault
```

Explanation:
In Path a the demarshal functions reinitialize the GValue to a type, that corresponds to the type of the GVariant, in this case a string.
(gchararray is a typedef for char *)
As such, the autofeeing works.

**Comment by muktupavels** [comment]
Your example just proves you are wrong... You used g_value_unset (correct thing to do) and after that you tried to free const variables that causes segfault that just shows that g_value_unset does required work - memory is freed.

To make it clear:

> BUT their data is not freed! Not even with g_value_unset(), as the GValues in the plugin only hold pointers to the data, which they don't know how to free.

But your output shows that values are freed after you use g_value_unset, at least you get segfault trying to free ids that was supposed to be const and should not be freed with g_strfreev.

**Comment by info-cppsp** [comment]
Yes.
Still, why haven't the original authors used the g_value_unset() ??
Note that the original g_free stuff also works - it has worked for years - it's just not that pretty...

Rewrote test with suggestions.
[test3.c.v2.txt](https://github.com/mate-desktop/mate-sensors-applet/files/1665658/test3.c.v2.txt)
Output:
```
model: 
dev: (null)
Segmentation fault
```
:(

**Comment by info-cppsp** [comment]
BUT then I was right the [first time](https://github.com/mate-desktop/mate-sensors-applet/pull/70#issuecomment-359813939).

**Comment by muktupavels** [comment]
Can not answer why original authors did that... But compiler is warning that you do something wrong...

Yes, it "works", but lets imagine that in future GValue gains ability to call free function automatically? Then you would get double free... Just because it worked it does not mean that it was correct. If something is returned as const it just means that you should not modify it and/or free it.

Not sure about your 'first time'. First you say that it is needed and then you say it is not needed in same comment...

My guess is that you can not go wrong if you use g_value_unset, but you can get memleaks not using it. So if not sure, just use g_value_unset and you wont get problems.

In other words - this merge proposal is correct fix.

**Comment by info-cppsp** [comment]
[test3.c.v3.txt](https://github.com/mate-desktop/mate-sensors-applet/files/1667382/test3.c.v3.txt)

Output:
```
model type name: gchararray
dev type name: gchararray
id type name: GStrv
model: Maxtor 6Y080L0
dev: /dev/sda
id: /dev/disk/by-id/ata-Maxtor_...
dummy: mydummystring

Printing strings from global variables:
model: Maxtor 6Y080L0
dev: /dev/sda
id: /dev/disk/by-id/ata-Maxtor_...
dummy: mydummystring
```

not freed :S

**Comment by muktupavels** [comment]
Right, because you are not using `g_value_unset`... You **must** do it!

**Comment by info-cppsp** [comment]
## Summary
1. As it turns out, contrary to my belief, there is no autofreeing for ALL GValues.
I somehow thought, that GValue come from (or is like) GObject, with which I have worked last.
(They normally free their properties...)

2. There IS autofreeing for int, float and pointer type GValues, as the value is stored inside the GValue structure and is freed at function end.

3. If  a GValue is of type G_TYPE_STRING (or variant or boxed), it (or better said its content) needs to be freed manually.
This can be done by
a, freeing the string directly (g_free())
OR
b, calling g_value_unset() on the GValue, which calls its freeing function.

There is NO freeing function for the boxed type GValue, unless you set one!

3. For the udisks plugin I think the original g_free code was used to provide clarity, that the strings are 100% freed.
(Also there may have been a version of dbus-glib, where the special case for handling GVariant DBus Message results didn't exist in dbus_g_proxy_call() or just gave a different result, like a boxed type GValue...)

4. Using g_value_unset() instead of the g_free functions works on an up-to-date mint 17.3.

5. We have found that dbus_g_proxy_call() has a special case for handling GVariant DBus Message results IF the user asks for a G_TYPE_VALUE. The demarshal functions in dbus-glib
a, reinitialize the GValue to a type based on the GVariant
b, copy value from GVariant to GValue
(c, likely GVariant is freed when the DBus Message is, but I didn't get into that)

6. G_TYPE_STRING doesn't work, bc [demarshal_basic()](http://maemo.org/api_refs/4.0/dbus-glib/dbus-gvalue_8c-source.html#l00446) gets called for it, which copies the value from the DBus Message result (in our case a GVariant) to the G_TYPE_STRING GValue.
BUT there is no case in demarshal_basic() for a GVariant, so all this fails.

Guy, sorry for all the drama and thanks for helping me understand GValues and dbus-glib better.
Too bad it was a [waste of time](https://github.com/mate-desktop/mate-sensors-applet/pull/72).

**Comment by muktupavels** [comment]
> 1. As it turns out, contrary to my belief, there is no autofreeing for ALL GValues.
> I somehow thought, that GValue come from (or is like) GObject, with which I have worked last.
> (They normally free their properties...)

> 2. There IS autofreeing for int, float and pointer type GValues, as the value is stored inside the GValue structure and is freed at function end.

I guess you are mixing stack and heap. `GValue` itself is allocated on stack. So variable is automatically deallocated. `G_TYPE_INT` all data is on stack. `G_TYPE_STRING` has pointer to memory on heap and it must be freed, that does not happen automatically. You **must** use `g_value_unset` to free that.

> If a GValue is of type G_TYPE_STRING (or variant or boxed), it (or better said its content) needs to be freed manually.
> This can be done by
> a, freeing the string directly (g_free())

No. Show me any book and/or anything else that recommends to free something that are returned as const. Show any other code that does it this way, I doubt you will find...

Look again - https://pastebin.com/yKr8UqKF to see what could happen if you free something you should not.

> OR
> b, calling g_value_unset() on the GValue, which calls its freeing function.

This is only correct way to do that. You may/can avoid `g_value_unset` if you know that type you use is fully on stack. If not sure always use `g_value_unset`.

> There is NO freeing function for the boxed type GValue, unless you set one!

It abstract type. See doc:
https://developer.gnome.org/gobject/stable/gobject-Generic-values.html#G-TYPE-VALUE:CAPS
> The type ID of the "GValue" type which is a boxed type, used to pass around pointers to GValues.

You get `G_TYPE_STRV` for ids. And it does have free function. Check this - https://pastebin.com/e84xjz4K and run under valgrind. You will see definitely and indirectly there is 0 bytes lost. Now comment out one or both g_value_unset and see what happens.

> 5. We have found that dbus_g_proxy_call() has a special case for handling GVariant DBus Message results IF the user asks for a G_TYPE_VALUE. The demarshal functions in dbus-glib
> a, reinitialize the GValue to a type based on the GVariant
> b, copy value from GVariant to GValue
> (c, likely GVariant is freed when the DBus Message is, but I didn't get into that)

> 6. G_TYPE_STRING doesn't work, bc demarshal_basic() gets called for it, which copies the value from the DBus Message result (in our case a GVariant) to the G_TYPE_STRING GValue.
> BUT there is no case in demarshal_basic() for a GVariant, so all this fails.

Yes, I was wrong on this. "Get" method has GVariant that is G_VARIANT_TYPE_VARIANT:
> The type of a box that contains any other value (including another variant).

So I guess `G_TYPE_VALUE` is something similar and is need to be able to return different types/data...

Anyway this is about build warnings and proposed fix is correct thing to do.

**Comment by muktupavels** [comment]
Last spam from me...

`g_debug("Added %s %s", path, id);`
It is accessing already freed id. This line should be moved above `g_value_unset` calls or `id` should be changed to `info->id`.

And as bonus you can do this to free memory automatically (no need for `g_value_unset`):
`g_auto(GValue) model_v = G_VALUE_INIT;`

Supported only on gcc and clang:
https://developer.gnome.org/glib/stable/glib-Miscellaneous-Macros.html#g-auto

**Comment by info-cppsp** [comment]
Thanks.
Autofree support is on the TODO.
